libbpf-tools/profile.c

```diff
@@ -0,0 +1,499 @@
+					} else {
+						for (i = 0; i < env.perf_max_stack_depth && ip[i]; i++) {
+							sym = syms__map_addr(syms, ip[i]);
+							printf(";%s", sym ? sym->name : "[unknown]");
+						}
+					}
+				}
+			}
+			if (!env.user_stacks_only) {
+				if (env.delimiter && next_key.user_stack_id >= 0)
+					printf(";-");
+				if (stack_id_err(next_key.kern_stack_id) ||
+				    (next_key.kern_stack_id >= 0 &&
+				     bpf_map_lookup_elem(sfd, &next_key.kern_stack_id, ip) != 0))
+					printf(";[Missed Kernel Stack]");
+				else {
+					for (i = 0; i < env.perf_max_stack_depth && ip[i]; i++) {
+						ksym = ksyms__map_addr(ksyms, ip[i]);
+						printf(";%s", ksym ? ksym->name : "[unknown]");
+					}
+				}
+			}
+			printf(" %lld\n", val);
+		} else {
+			// print default multi-line stack output
+			if (!env.user_stacks_only) {
+				if (stack_id_err(next_key.kern_stack_id) ||
+				    (next_key.kern_stack_id >= 0 &&
+				     bpf_map_lookup_elem(sfd, &next_key.kern_stack_id, ip) != 0))
+					printf("    [Missed Kernel Stack]\n");
+				else {
+					for (i = 0; i < env.perf_max_stack_depth && ip[i]; i++) {
+						ksym = ksyms__map_addr(ksyms, ip[i]);
+						printf("    %s\n", ksym ? ksym->name : "[unknown]");
+					}
+				}
+			}
+
+			if (!env.kernel_stacks_only) {
+				if (env.delimiter && next_key.kern_stack_id >= 0)
+					printf("    --\n");
+				if (stack_id_err(next_key.user_stack_id) ||
+				    (next_key.kern_stack_id >= 0 &&
+				     bpf_map_lookup_elem(sfd, &next_key.user_stack_id, ip) != 0))
+					printf("    [Missed User Stack]\n");
+				else {
+					syms = syms_cache__get_syms(syms_cache, next_key.pid);
+					if (!syms) {
+						fprintf(stderr, "failed to get syms\n");
+					} else {
+						for (i = 0; i < env.perf_max_stack_depth && ip[i]; i++) {
+							sym = syms__map_addr(syms, ip[i]);
+							printf("    %s\n", sym ? sym->name : "[unknown]");
+						}
+					}
+				}
+			}
+
+			printf("    %-16s %s (%d)\n", "-", next_key.name, next_key.pid);
+			printf("        %lld\n\n", val);
+		}
+	}
+
+	if (missing_stacks > 0) {
+		fprintf(stderr, "WARNING: %d stack traces could not be displayed.%s\n",
+			missing_stacks, has_collision ?
+			" Consider increasing --stack-storage-size.":"");
+	}
+
+cleanup:
+	free(ip);
+}
+
+int main(int argc, char **argv)
+{
+	static const struct argp argp = {
+		.options = opts,
+		.parser = parse_arg,
+		.doc = argp_program_doc,
+	};
+	struct syms_cache *syms_cache = NULL;
+	struct ksyms *ksyms = NULL;
+	struct bpf_link *links[MAX_CPU_NR] = {};
+	struct profile_bpf *obj;
+	int err, i;
+	char* stack_context = "user + kernel";
+	char thread_context[64];
+	char sample_context[64];
+
+	err = argp_parse(&argp, argc, argv, 0, NULL, NULL);
+	if (err)
+		return err;
+	if (env.user_stacks_only && env.kernel_stacks_only) {
+		fprintf(stderr, "user_stacks_only and kernel_stacks_only cannot be used together.\n");
+		return 1;
+	}
+
+	libbpf_set_print(libbpf_print_fn);
+
+	err = bump_memlock_rlimit();
```

Comment:
Please refer to other tools, we don't need this anymore.

Reply:
I fixed it. Thank you.